feat: add total reward for epoch for each user

neocybereth opened the pull request.

Fraccaman commented:
Ehi! you are still missing the migration, you need to add a folder here `https://github.com/anoma/namada-indexer/tree/main/orm/migrations` (you can follow the guide `https://diesel.rs/guides/getting-started#setup-diesel-for-your-project` to auto generate the folder for you). The migration is needed to modify the rewards table schema and add the `epoch` field

neocybereth commented:
Oof! Sorry about that, coming right up!

joel-u410 commented on orm/migrations/2024-12-16-231256_add_epoch_pos_rewards_table/up.sql:
```diff
@@ -0,0 +1,3 @@
+-- Alters the pos_rewards table to add a fourth column:
+-- epoch: INTEGER, representing the epoch number
+ALTER TABLE pos_rewards ADD COLUMN epoch INTEGER NOT NULL;
```
This migration will fail if there is any existing data.

You could do something like the following:

```
-- Populate existing records with epoch = 0
ALTER TABLE pos_rewards ADD COLUMN epoch INTEGER NOT NULL DEFAULT 0;

-- Now we can safely drop the default
ALTER TABLE pos_rewards ALTER COLUMN epoch DROP DEFAULT;
```

joel-u410 commented on rewards/src/schema.rs:
```diff
@@ -0,0 +1,297 @@
+// @generated automatically by Diesel CLI.
```
I think this file belongs in `orm/src/schema.rs` doesn't it? (And it's already there -- so probably this one should not be added)

neocybereth replied:
Updated and removed

joel-u410 commented on shared/src/balance.rs:
```diff
@@ -35,6 +35,12 @@ impl Display for Amount {
 }
 
 impl Amount {
+    pub fn as_i64(&self) -> i64 {
+        let s = self.0.to_string();
+        s.parse::<i64>()
+            .expect("Cannot convert NamadaAmount to i64")
+    }
```
I'm not sure `as_i64` is idiomatic Rust -- you might instead implement the `From` trait here, and you might as well go directly to `BigDecimal` too.

It would look something like this:
```rust
use std::str::FromStr;

impl From<Amount> for BigDecimal {
    fn from(amount: Amount) -> BigDecimal {
        BigDecimal::from_str(&amount.0.to_string_native())
            .expect("Invalid amount")
    }
}
```

and then to use it you'd just do:
```rust
            raw_amount: BigDecimal::from(reward.amount),
```

I've implemented this idea in [this commit](https://github.com/anoma/namada-indexer/commit/8eff9a20817c402920f17fce0b5f1bf1111a4cba) (along with some tests for the conversions) if you'd like to have a look.

neocybereth replied:
Updated

Fraccaman commented:
you might also want to update the webserver? I mean, update/create a new endpoint

joel-u410 commented on orm/src/schema.rs:
```diff
@@ -227,6 +183,7 @@ diesel::table! {
         owner -> Varchar,
         validator_id -> Int4,
         raw_amount -> Numeric,
+        epoch -> Int4,
```
Hi @neocybereth -- I have another suggestion here, it looks like you've reformatted a bunch of things in this file that aren't really part of this branch. It would be best to keep your changes minimal and not include these formatting changes.

I _think_ this line is the only change that you should actually have in this file.

joel-u410 replied:
I just pushed a commit that restores that formatting here, and also has one other small refactor in the `rewards/src/services/namada.rs` to reuse the existing function for querying current epoch instead of copying it: https://github.com/anoma/namada-indexer/commit/2d0a5ac44b660ea656505158b4232e6493db3486

Feel free to incorporate it into your branch here

neocybereth replied:
alright I've integrated both of those 👍

joel-u410 commented on orm/migrations/2024-12-16-231256_add_epoch_pos_rewards_table/up.sql:
```diff
@@ -0,0 +1,6 @@
+-- Alters the pos_rewards table to add a fourth column:
+-- epoch: INTEGER, representing the epoch number
+-- Populate existing records with epoch = 0
+ALTER TABLE pos_rewards ADD COLUMN epoch INTEGER NOT NULL DEFAULT 0;
+-- Now we can safely drop the default
+ALTER TABLE pos_rewards ALTER COLUMN epoch DROP DEFAULT;
```
While testing this, I ran into a problem where the `pos_rewards` table didn't have an updated `UNIQUE` constraint which incorporates the `epoch` column.

I pushed a commit that fixes that: https://github.com/anoma/namada-indexer/commit/725f9c9255038f55656be5ab2a6f695c928dbb6e

Feel free to add it to your branch also.

neocybereth replied:
Updated that here:

https://github.com/anoma/namada-indexer/pull/204/commits/1244bab596a67465b933deea590b7a291ad3b5e8

joel-u410 commented on orm/src/pos_rewards.rs:
```diff
@@ -1,29 +1,26 @@
-            raw_amount: BigDecimal::from_str(&reward.amount.to_string())
-                .expect("Invalid amount"),
             validator_id,
+            raw_amount: BigDecimal::from(reward.amount),
```
Note - this is assigning to `raw_amount` the _decimal_ value of `nam` (i.e. `12.34567` not `1234567 unam`) -- which is different than the previous behavior of the `rewards` crawler. I noticed it in my testing because the `raw_amount` column has type `numeric(78,0)` which only stores integer values, and the reward amounts were being truncated.

neocybereth replied:
https://github.com/anoma/namada-indexer/pull/204/commits/0957a4a6e60f30b9048b47797379fb12b978b8ab

Fixed here with: 

```
            raw_amount: BigDecimal::from_str(&reward.amount.to_string()).unwrap(),
```

neocybereth commented:
> you might also want to update the webserver? I mean, update/create a new endpoint

https://github.com/anoma/namada-indexer/pull/204/commits/fae5e1fbdcb7b96b4212ebbceec1def437623858 all the webserver endpoints are here. 

basically it takes a delegator address, validator address and epoch and gives back the rewards. The validator address is used to find a validator id to look up the reward. Again, my rust sucks so let me know if there's anything off.

brentstone commented:
Does this depend on https://github.com/anoma/namada/pull/4196?

neocybereth commented:
> Does this PR depend on [anoma/namada#4196](https://github.com/anoma/namada/pull/4196)?

I'm 99% sure yes, but I'm also a noob to this codebase so @joel-u410 or @Fraccaman could confirm for sure.

joel-u410 commented:
> Does this PR depend on https://github.com/anoma/namada/pull/4196?

Hi @brentstone @neocybereth -- no, currently it does not, but after https://github.com/anoma/namada/pull/4196 is merged I would like this one to be updated to take advantage of it.

Right now, in this branch, it only queries "current epoch" rewards, and just assigns the current epoch number to their records -- after https://github.com/anoma/namada/pull/4196 is merged I would want to change this to use the `epoch` argument to be more precise about which epoch is being requested when rewards are indexed.

joel-u410 commented:
Just realized something here ... [delete_claimed_rewards](https://github.com/anoma/namada-indexer/blob/455cf5a18236545a5a25af5fc5db02b113c34390/chain/src/repository/pos.rs#L197) is now going to delete things it shouldn't.

Instead of deleting, it should probably insert rows with `raw_amount = 0` to indicate they've been claimed.

joel-u410 commented:
@neocybereth I'm looking at the existing `/pos/reward/:address` endpoint and it also needs to be updated to return only the _latest_ reward -- right now it's returning every reward from every epoch in a giant list 😁

neocybereth commented:
@joel-u410 epic, tahnk you for the heads up! What's your go to testing methodology here? I'm just shooting in the dark and YOLO-ing atm and surely there's gotta be a better way :D

joel-u410 commented:
> What's your go to testing methodology here?

I'm not specifically testing this PR, but I am _using_ it in production right now -- so I suppose that's the best testing of all :)

And, so I wasn't testing this per se, but I was working on some balance queries and noticed that I was getting way too many `rewards` entries in the response. So I started investigating the code more closely. And while reading through the code I also just happened to notice the `delete_claimed_rewards` issue too.

neocybereth commented:
> Just realized something here ... [delete_claimed_rewards](https://github.com/anoma/namada-indexer/blob/455cf5a18236545a5a25af5fc5db02b113c34390/chain/src/repository/pos.rs#L197) is now going to delete things it shouldn't.
> 
> Instead of deleting, it should probably insert rows with `raw_amount = 0` to indicate they've been claimed.
> 
> Which will need to be distinguished somehow from the `pos_rewards` entry in the same `epoch`.
> 
> We'll probably need to add a column -- either `block_height` perhaps (to indicate the block where claim occurred) or maybe just a boolean `claimed` column would work.

Alrighty commits `f2aade57b2a0295b696c724770dc098457450cd0` and `abd16ec6785bb48c6032d762bf45ebc991c53490` fix this.  I chose the `claimed` route

neocybereth commented:
> @neocybereth I'm looking at the existing `/pos/reward/:address` endpoint and it also needs to be updated to return only the _latest_ reward -- right now it's returning every reward from every epoch in a giant list 😁

resolved this one here: https://github.com/anoma/namada-indexer/pull/204/commits/b1296555e1c0409a23dd786b11d372828f0a764e

neocybereth commented:
> > Does this PR depend on [anoma/namada#4196](https://github.com/anoma/namada/pull/4196)?
> 
> Hi @brentstone @neocybereth -- no, currently it does not, but after [anoma/namada#4196](https://github.com/anoma/namada/pull/4196) is merged I would like this one to be updated to take advantage of it.
> 
> Right now, in this branch, it only queries "current epoch" rewards, and just assigns the current epoch number to their records -- after [anoma/namada#4196](https://github.com/anoma/namada/pull/4196) is merged I would want to change this to use the `epoch` argument to be more precise about which epoch is being requested when rewards are indexed.
> 
> (I have [a Draft commit](https://github.com/anoma/namada-indexer/pull/235/commits/fbcb80b63235af85ce288cdd41e491ba0d0ceadf) where I've started working on that, including adding support for `--backfill-from` to `rewards` -- see #193 for context on the `--backfill-from` feature)

@joel-u410 @brentstone any idea when [anoma/namada#4196] will get merged in?

joel-u410 commented:
> resolved this one here: [b129655](https://github.com/anoma/namada-indexer/commit/b1296555e1c0409a23dd786b11d372828f0a764e)

In my fork, for now, I've added [this commit](https://github.com/anoma/namada-indexer/commit/02ada1263c8e07f7e14fd323189307aa4c34031f) which fixes the `/pos/reward/:address` API route to restore its original behavior of returning only the current/latest `reward` entry.

I don't think your commit will work unless you also change the route in `webserver/src/app.rs` to be `/pos/reward/:address/:epoch` -- but that would be a breaking change to the API, and would require the caller to specify an epoch (which in most cases they probably won't want to do).

Perhaps though a 2nd endpoint, or _optional_ `epoch` parameter, would be useful for (optionally) querying for past epochs. But I think the default should remain as "query latest epoch" if the user doesn't specify.

neocybereth commented:
> > resolved this one here: [b129655](https://github.com/anoma/namada-indexer/commit/b1296555e1c0409a23dd786b11d372828f0a764e)
> 
> In my fork, for now, I've added [this commit](https://github.com/anoma/namada-indexer/commit/02ada1263c8e07f7e14fd323189307aa4c34031f) which fixes the `/pos/reward/:address` API route to restore its original behavior of returning only the current/latest `reward` entry.
> 
> I don't think your commit will work unless you also change the route in `webserver/src/app.rs` to be `/pos/reward/:address/:epoch` -- but that would be a breaking change to the API, and would require the caller to specify an epoch (which in most cases they probably won't want to do).
> 
> Perhaps though a 2nd endpoint, or _optional_ `epoch` parameter, would be useful for (optionally) querying for past epochs. But I think the default should remain as "query latest epoch" if the user doesn't specify.

Nice, I updated that here: https://github.com/anoma/namada-indexer/pull/204/commits/d36ec9e04da4ff657fa1a9a4eddf33116534497b to match your commit

Fraccaman commented:
This looks very cool! I was wondering, do you think we could add a `epoch` query parameters to the `v1/pos/reward/<address>` route ?

joel-u410 commented:
> This looks very cool! I was wondering, do you think we could add a `epoch` query parameter to the `v1/pos/reward/<address>` route ?

@Fraccaman I took a shot at adding that in [this commit](https://github.com/anoma/namada-indexer/commit/68d63cf0aae052e7ee0a93bc7d4b65a4b2681bf1). I haven't tested it however.

If this looks good, @neocybereth feel free to cherry-pick it into your branch here.

neocybereth commented:
Sounds good. Let me know if that's all good @Fraccaman and I'll cherry pick it in.

neocybereth commented:
Alright I added in https://github.com/anoma/namada-indexer/commit/68d63cf0aae052e7ee0a93bc7d4b65a4b2681bf1
in my latest commit https://github.com/anoma/namada-indexer/commit/7320744c639297d51f41821cb0c89d8f0c684005

neocybereth commented:
@joel-u410 I guess now that https://github.com/anoma/namada/pull/4196 is merged this one needs to be updated to take advantage of it like you mentioned above?

If so, what in particular do you want done to this PR to update it accordingly?

joel-u410 commented:
> @joel-u410 I guess now that [anoma/namada#4196](https://github.com/anoma/namada/pull/4196) is merged this one needs to be updated to take advantage of it like you mentioned above?
> 
> If so, what in particular do you want done to this PR to update it accordingly?

I think this should be merged as-is, then we can follow up later to update it with the new functionality from https://github.com/anoma/namada/pull/4196. That requires nodes to be updated in order to take advantage, so we'll probably need to wait till the next `namada` release for it anyway.

neocybereth commented:
Roger that. @Fraccaman this is ready for a review when you next get a chance 🙏

Fraccaman commented:
I think this needs to be rebased `1.0.0-maint`

neocybereth commented:
Alright I rebased to `1.0.0-maint` but it doesn't seem right to me. @joel-u410 the changes left when comparign to 1.0.0-maint after rebase seem like many more files than I've touched, were there any additionals that you're seeing here that seem out of whack as well?

Fraccaman commented:
something is messed up in here :S

neocybereth commented:
> something is messed up in here :S

Yeaaaaa it was my rebase. I'm going to have to go through and piece the PR back together.

sug0 commented:
do you still have the old tip of your branch in your `git reflog`? it would be easier to take that as a starting point

neocybereth commented:
> do you still have the old tip of your branch in your `git reflog`? it would be easier to take that as a starting point

omg you're a lifesaver!

neocybereth commented:
@Fraccaman okay, we're back to clean. Everythings up-to-date against `main`.
 
Anything else I need to do before getting approval? Shall I attempt another (hopefully successful :D) rebase against `1.0.0-maint` from here?

Fraccaman commented:
@neocybereth can you take a look at. theci failures?

neocybereth commented:
@Fraccaman sorry I've been busy with other work! I've fixed up all the CI issues.

sug0 commented on webserver/src/repository/pos.rs:
```diff
@@ -325,11 +355,21 @@ impl PosRepositoryTrait for PosRepository {
+                pos_rewards::table
+                    .select(diesel::dsl::max(pos_rewards::epoch))
+                    .first::<Option<_>>(conn)
+                    .unwrap_or(Some(0))
+                    .unwrap_or(0)
```
some of these will be legitimate database errors, better to propagate the errors if possible

sug0 commented on webserver/src/service/pos.rs:
```diff
@@ -302,6 +304,51 @@ impl PosService {
+        let db_validator = match db_validator {
+            Some(val) => val,
+            None => {
+                tracing::error!("No validator found for the given address");
```
this `tracing::error!()` will pollute the logs of the webserver for no good reason (it doesn't give any meaningful information about the failure). let's change it to a debug log and include the address of the queried validator.

```suggestion
                tracing::debug!(%validator_address, "No validator found for the given address");
```

sug0 commented on webserver/src/service/pos.rs:
```diff
@@ -302,6 +304,51 @@ impl PosService {
+        // 1) Lookup validator by its address
+        let db_validator = self
+            .pos_repo
+            .find_validator_by_address(validator_address)
```
```suggestion
            .find_validator_by_address(&validator_address)
```

neocybereth commented:
@Fraccaman @sug0 are these good to merge?

neocybereth commented:
@Fraccaman @sug0 just checking back in to see if these are good to merge?
